### ISSUE TITLE
Encoder random hang

### ISSUE THREAD

**Issue by tianjunwork** [OpenVisualCloud/SVT-HEVC#497]
> `./SvtHevcEncApp -i ../../../../yuv/bbb_1920x1080_420p.yuv -w 1920 -h 1080 -encMode 0 -intra-period 100 -scd 0 -rc 0 -q 22 -n 1000 -b 0.265` random hang.
> I wonder if it can be fixed by this PR too.

Thanks for capturing this!

At first, this is another (fixed, not random) hang issue which hasn't been addressed by this PR, or isn't regression caused by this PR.

And here is the hang pattern due to encMode 0 and intra period:
- ./SvtHevcEncApp -i bbb_1920x1080_420p.yuv -w 1920 -h 1080 **-encMode 0 -intra-period 100** -scd 0 -rc 0 -q 22 -n 1000 -b 0.265
  - Hangs after fixed encoded frames 253
- ./SvtHevcEncApp -i /home/austin/VCD/bbb_1920x1080_420p.yuv -w 1920 -h 1080 **-intra-period 100** -scd 0 -rc 0 -q 22 -n 1000 -b 0.265
  - Hangs after fixed encoded frames 286
- ./SvtHevcEncApp -i /home/austin/VCD/bbb_1920x1080_420p.yuv -w 1920 -h 1080 **-intra-period 55** -scd 0 -rc 0 -q 22 -n 1000 -b 0.265
  - Works fine (-intra-period is [recommended](https://github.com/OpenVisualCloud/SVT-HEVC/blob/master/Docs/svt-hevc_encoder_user_guide.md) to set as 55 with 60 FPS encoded)
- ./SvtHevcEncApp -i /home/austin/VCD/bbb_1920x1080_420p.yuv -w 1920 -h 1080 **-encMode 0** -scd 0 -rc 0 -q 22 -n 1000 -b 0.265
  - Hangs (only for encMode 0) after fixed encoded frames 283

Anyway, it's another hang issue we need to look at.

_Originally posted by @Austin-Hu in https://github.com/OpenVisualCloud/SVT-HEVC/pull/475#issuecomment-597735296_

**Comment by Austin-Hu** [comment]
The hang issue can be 100% reproducible with command:
```
./SvtHevcEncApp -i ~/bbb_1920x1080_420p.yuv -w 1920 -h 1080 -encMode 0 -intra-period 100 -scd 0 -rc 0 -q 22 -n 1000 -b 0.265
```
with v1.3.0 or v1.4.0 tag, after encoding 295 frames.

**Comment by tianjunwork** [comment]
On current master tip(Disable OUT_ALLOC), encoder **randomly** hang with only with `-encMode 0`, not other encMode.
`./SvtHevcEncApp -i ../../../yuv/bbb_1920x1080_420p.yuv -w 1920 -h 1080 -encMode 0 -n 1000`
Hang seems has nothing to do with `-intra-period 100`.
And I can't reproduce hang with `./SvtHevcEncApp -i bbb_1920x1080_420p.yuv -w 1920 -h 1080 -intra-period 100 -scd 0 -rc 0 -q 22 -n 1000 -b 0.265`

**Comment by tianjunwork** [comment]
Did regression test with `./SvtHevcEncApp -i ../../../yuv/bbb_1920x1080_420p.yuv -w 1920 -h 1080 -encMode 0 -n 1000`. It is easier to reproduce on v1.3, nearly 100%.
```
(v1.3) $ ./100run.sh
-------------------------------------------- run: 1
-------------------------------------------
SVT-HEVC Encoder
SVT [version]:  SVT-HEVC Encoder Lib v1.3.0
SVT [build]  :  GCC 8.1.0        64 bit
LIB Build date: Mar 17 2020 22:11:42
-------------------------------------------
Number of logical cores available: 72
Number of PPCS 107
-------------------------------------------
SVT [config]: Main10 Profile    Tier (auto)     Level (auto)
SVT [config]: EncoderMode / LatencyMode / Tune                                  : 0 / 0 / 1
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1920 / 1080
SVT [config]: FrameRate / Gop Size                                              : 60 / 101
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 3 / 0 / 2
SVT [config]: BRC Mode / QP  / LookaheadDistance / SceneChange                  : CQP / 32 / 17 / 1
-------------------------------------------
Encoding       296
./100run.sh: line 14:  6326 Killed                  ./SvtHevcEncApp -i ../../../../yuv/bbb_1920x1080_420p.yuv -w 1920 -h 1080 -encMode 0 -intra-period 100 -n 1000
-------------------------------------------- run: 2
-------------------------------------------
SVT-HEVC Encoder
SVT [version]:  SVT-HEVC Encoder Lib v1.3.0
SVT [build]  :  GCC 8.1.0        64 bit
LIB Build date: Mar 17 2020 22:11:42
-------------------------------------------
Number of logical cores available: 72
Number of PPCS 107
-------------------------------------------
SVT [config]: Main10 Profile    Tier (auto)     Level (auto)
SVT [config]: EncoderMode / LatencyMode / Tune                                  : 0 / 0 / 1
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1920 / 1080
SVT [config]: FrameRate / Gop Size                                              : 60 / 101
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 3 / 0 / 2
SVT [config]: BRC Mode / QP  / LookaheadDistance / SceneChange                  : CQP / 32 / 17 / 1
-------------------------------------------
Encoding      1000
SUMMARY --------------------------------- Channel 1  --------------------------------
Total Frames            Frame Rate              Byte Count              Bitrate
        1000            60.00 fps                  1288499              618.48 kbps


Channel 1
Average Speed:          9.14 fps
Total Encoding Time:    109434 ms
Total Execution Time:   110870 ms
Average Latency:        12093 ms
Max Latency:            23545 ms
Encoder finished
-------------------------------------------- run: 3
-------------------------------------------
SVT-HEVC Encoder
SVT [version]:  SVT-HEVC Encoder Lib v1.3.0
SVT [build]  :  GCC 8.1.0        64 bit
LIB Build date: Mar 17 2020 22:11:42
-------------------------------------------
Number of logical cores available: 72
Number of PPCS 107
-------------------------------------------
SVT [config]: Main10 Profile    Tier (auto)     Level (auto)
SVT [config]: EncoderMode / LatencyMode / Tune                                  : 0 / 0 / 1
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1920 / 1080
SVT [config]: FrameRate / Gop Size                                              : 60 / 101
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 3 / 0 / 2
SVT [config]: BRC Mode / QP  / LookaheadDistance / SceneChange                  : CQP / 32 / 17 / 1
-------------------------------------------
Encoding       295
./100run.sh: line 14:  7689 Killed                  ./SvtHevcEncApp -i ../../../../yuv/bbb_1920x1080_420p.yuv -w 1920 -h 1080 -encMode 0 -intra-period 100 -n 1000
```

**Comment by tianjunwork** [comment]
Equivalent ffmpeg command also hang: `./ffmpeg -stream_loop -1 -video_size 1920x1080 -r 60 -i ../../yuv/bbb_1920x1080_420p.yuv -c:v libsvt_hevc -preset 0 -y 0.mp4`

**Comment by Austin-Hu** [comment]
**Root cause of this issue:**
The single thread SvtHevcEncApp (and other single thread encoding frameworks such as ffmpeg) calls EbH265EncSendPicture, EbH265GetPacket and EbH265ReleaseOutBuffer sequentially. But EbH265EncSendPicture and EbH265GetPacket are asynchronous, so that application would get the 1st available encoded packet after sending several frames. When EbH265EncSendPicture hangs due to exhausted Input FIFO, application couldn't release Output FIFO in time and the pending frames in the encoding pipeline are also blocked in the single thread InitialRateControl kernel where Output FIFO objects are dequeued. So the application and encoder are locked each other.

**Root cause tracing details:**
1. Application repeat sends frames (until hangs at the 519th frame);
2. Because it takes random time to complete encoding a frame, the application would randomly get the encoded packets. When encoding hangs, application would get totally **295** (or **296**, **297** or **298**) frames, and release Output objects with corresponding times;
3. During encoding, the stages (through EncDec -> Packetization -> RaceControl (with releasing Input objects)) complete for 405 times. Right now application is sending the 519th frame (519 - 405 = 114). But the Input FIFO just has 113 objects, so that **application hangs at EbH265EncSendPicture**, cand can't go ahead to call following EbH265ReleaseOutBuffer to release Output object;
4. But currently all the (totally 113) Output FIFO objects have been exhausted. At the moment the single thread InitialRateControl kernel gets POC 412, it will handle POC 395~412 (in group of LookAheadDistance = 17 as sliding window). But when it handles POC **408** (or **409**, **410** or **411**), where **408 + 1 - 295 = 114**, it can't dequeue any object from the Output FIFO, so that the **InitialRateControl kernel hangs**;

(BTW, moving the procedure of dequeuing Output FIFO objects from InitialRateControl to Packetization couldn't help resolve this issue. It also hangs when EncDec feeds back POC 262 to PictureManger, but the single thread PictureManger has already hung at dequeuing the (totally 18) Child PCS FIFO objects, and couldn't handle the posted POC 262 from EncDec.)

**Options to address this issue:**
1. Decouple EbH265EncSendPicture and EbH265GetPacket calling into different threads in encoding applications (and frameworks), to work around this issue, like what PR #507 does;
2. Calculate the correct number of each FIFO objects, passing through some kernels or the whole encoding pipeline, depending on the specific encoding FPS and time consumed (during runtime) in each encoding stage. So each FIFO have enough objects to be used, at any time and anywhere.
3. Increase the number of some FIFO objects (such as Input, Output, etc.) with some tune effort.

**Comment by tianjunwork** [comment]
Fixed by #523
